### PR TITLE
feat: send webhooks to Salient CompleteView events

### PR DESCRIPTION
**Pull request by danleyb2**
https://trello.com/c/r4gqkBUA/1307-salient-vms-integration

### CODE REVIEW

**Review by marcbelmont — CHANGES_REQUESTED** [review]
- [x] move to webhooks
- [ ] merge Salient and NX

**Comment by marcbelmont** [comment]
@danleyb2 
Don't forget to work on this.

File: webhooks/forwarder/destinations/salient.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            )
+            res.raise_for_status()
+        except requests.exceptions.HTTPError as errh:
+            lgr.error("Http Error:", errh)
+        except requests.exceptions.ConnectionError as errc:
+            lgr.error("Error Connecting:", errc)
+        except requests.exceptions.Timeout as errt:
+            lgr.error("Timeout Error:", errt)
+        except requests.exceptions.RequestException as err:
+            lgr.error("OOps: Something Else", err)
```
Comment (adolfoarmas) [inline webhooks/forwarder/destinations/salient.py:53]: NIT:
```suggestion
            lgr.error("Oops: Something Else", err)
```

File: webhooks/forwarder/destinations/nx.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            )
+            res.raise_for_status()
+        except requests.exceptions.HTTPError as errh:
+            lgr.error("Http Error:", errh)
+        except requests.exceptions.ConnectionError as errc:
+            lgr.error("Error Connecting:", errc)
+        except requests.exceptions.Timeout as errt:
+            lgr.error("Timeout Error:", errt)
+        except requests.exceptions.RequestException as err:
+            lgr.error("OOps: Something Else", err)
```
Comment (adolfoarmas) [inline webhooks/forwarder/destinations/nx.py:49]: NIT:
```suggestion
            lgr.error("Oops: Something Else", err)
```

Comment (adolfoarmas) [inline webhooks/forwarder/destinations/nx.py:49]: Kindly confirm:
- Which exception collects the metadata "camera" name error?
- Is the error coming from the API explanatory enough regarding the wrong camera name for example?
- Same questions for Salient deviceGuid.

**Review by adolfoarmas — COMMENTED** [review]
Regarding the folder organization:

I propose the organization on webhooks VMS to be:

webhooks -> VMS -> 
- NX_or_Salient _(this forwarder)_
- OpenEye
- Synology

That way all VMS/VMM integrations are bundled in one folder.

**Comment by danleyb2** [comment]
@adolfoarmas regarding the folder structure, The larger idea is to also add OpenEye and Synology as a destination in the **forwarder**. This is because they are all basically a server that format the webhook payload then call and external API

**Comment by ttayson** [comment]
@marcbelmont @danleyb2 
I have only one point to highlight: I suggest the creation of bookmarks instead of events, along with the use of the latest NX API that does not require Digest authentication. The use of bookmarks makes the boards searchable in the system, in addition to providing a newer and recommended authentication method.

We have two NX implementations in our guides. The implementation below utilizes bookmarks:
https://github.com/parkpow/deep-license-plate-recognition/tree/master/webhooks/Webhook_nx

Is there any limitation for using bookmarks that made us choose events instead?

**Comment by danleyb2** [comment]
@ttayson @marcbelmont  @adolfoarmas 
Suggestions:
- [ ]  Keep existing REST API **NX implementation**  and delete the one in PR
- [ ]  Maintain the implementations in separate folders and so refactor this PR to only contain Salient

Question:
How should we handle support for different NX VMS versions because the endpoint changes? It's easier and recommended to support latest only then advise users to upgrade 

Based on **NX API versions and their support** from [here](https://meta.nxvms.com/doc/developers/api-tool/main?type=1&system=7)
_If the integration intends to support VMS 5.0+, it must prefer `/rest/v1` which was introduced in this VMS version. 
If it intends to support VMS 5.1+, it must prefer `/rest/v2`, and so on. 
Originally (up to and including VMS 4.2), there was a mostly procedural-style API, with URLs mostly starting with `/api/` and `/ec2/` - these functions are now referred to as a "LEGACY & BETA" API.
Latest is `/rest/v3`_

File: webhooks/webhook_salient/main.py [hunk 3]
```diff
@@ -27,23 +27,30 @@
...
-                "description": description,
+            json={
+                "events": [
+                    {
+                        "entityType": 1,
+                        "eventType": 58,
+                        "eventDescription": f"Plate Detection [{description}]",
+                        "user": f"Platerecognizer({source})",
+                        "deviceGuid": camera_uid,
+                        # "streamId": 0
```
Comment (adolfoarmas) [inline webhooks/webhook_salient/main.py:49]: NIT: remove if not needed.

```suggestion
```

File: webhooks/webhook_salient/main.py [hunk 4]
```diff
@@ -122,13 +129,13 @@ def do_POST(self):
...
     parser = argparse.ArgumentParser(
-        description="Forward Stream Webhook Events to NX VMS as Alerts."
+        description="Forward Stream Webhook Events to CompleteView VMS as Events."
     )
-    parser.add_argument("--username", help="NX VMS Username.", required=True)
-    parser.add_argument("--password", help="NX VMS Password.", required=True)
-    parser.add_argument("--vms", help="VMS API Endpoint.", required=True)
+    parser.add_argument("--username", help="Recording Server Username.", required=True)
+    parser.add_argument("--password", help="Recording Server Password.", required=True)
+    parser.add_argument("--vms", help="Recording Server API Endpoint.", required=True)
```
Comment (adolfoarmas) [inline webhooks/webhook_salient/main.py:136]:
```suggestion
    parser.add_argument("--username", help="Salient VMS Username.", required=True)
    parser.add_argument("--password", help="Salient VMS Password.", required=True)
    parser.add_argument("--vms_api_url", "-v" help="Salient VMS API Endpoint.", required=True)
```

The `vms` variable name will have to be changed in other lines of code.

File: webhooks/webhook_salient/README.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+2. Run Image
+    ```bash
+
+    docker run --rm -t \
+       --net=host \
+       -e LOGGING=DEBUG \  # Turn on debug logging
+       platerecognizer/stream-salient-notifier \
+       --username=user \  # Recording server username
+       --password=pass \  # Recording server password
+       --vms='http://localhost:4502' \  # Recording server API Endpoint
```
Comment (adolfoarmas) [inline webhooks/webhook_salient/README.md:20]:
```suggestion
       --vms_api_url='http://localhost:4502' \  # Recording server API Endpoint
```

File: webhooks/webhook_salient/README.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+
+    # Example
+    docker run --rm -t \
+       --net=host \
+       -e LOGGING=DEBUG \
+       platerecognizer/stream-salient-notifier \
+       --username=admin \
+       --password=39393jdhhdiisu2 \
+       --vms='http://192.168.100.6:4502' \
```
Comment (adolfoarmas) [inline webhooks/webhook_salient/README.md:31]:
```suggestion
       --vms_api_url='http://192.168.100.6:4502' \
```

**Review by marcbelmont — APPROVED** [review]

**Pull request merged**